app/models/stripe_record.rb

```diff
@@ -36,28 +49,33 @@ class StripeRecord < ApplicationRecord
   # Also saves us from some pains because JSON columns are highly inconsistent among MySQL and MariaDB.
   serialize :parameters, coder: JSON
 
+  def determine_wca_status
+    result = WCA_TO_STRIPE_STATUS_MAP.find { |key, values| values.include?(stripe_status) }
+    result&.first || raise("No associated wca_status for stripe_status: #{stripe_status} - our tests should prevent this from happening!")
+  end
+
   def find_account_id
     self.account_id || parent_transaction&.find_account_id
   end
 
-  def update_status(api_transaction)
+  def assign_status(api_transaction)
     stripe_error = nil
 
-    case self.api_type
+    case self.stripe_record_type
     when 'payment_intent'
       stripe_error = api_transaction.last_payment_error&.code
     when 'charge'
       stripe_error = api_transaction.failure_message
     end
 
-    self.update!(
-      status: api_transaction.status,
+    self.assign_attributes(
```

Comment:
Be cautious here! Previously it was `update!` which triggers a DB write (and which is also intended to trigger a DB write here).
Your new `assign_attributes` does not trigger a DB write!

Reply:
This was intentional - we can't update the status at this point, as we now have a validation on the confirmation/canceled_at as well. So we either have to
- also set confirmed/canceled at as part of this function
- delay the db write until confirmed/canceled_at are set in the calling function

I prefer the latter, as it doesn't feel appropriate/intuitive for a stripe_record method to be updating properties of the payment_intent object